cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -28,18 +34,137 @@
 public class BasicWorkflowToken implements WorkflowToken {
   private Map<String, Map<String, Long>> mapReduceCounters;
 
+  private Map<String, List<NodeValueEntry>> tokenValueMap = Maps.newHashMap();
+
+  private String nodeName;
+
+  void setCurrentNode(String nodeName) {
+    this.nodeName = nodeName;
+  }
+
+  /**
+   * Merge the other WorkflowToken passed to the method as a parameter
+   * with the WorkflowToken on which the method is invoked.
+   * @param other the other WorkflowToken to be merged
+   */
+  void mergeToken(BasicWorkflowToken other) {
+    Map<String, List<NodeValueEntry>> otherTokenValueMap = other.tokenValueMap;
+
+    for (Map.Entry<String, List<NodeValueEntry>> entry : otherTokenValueMap.entrySet()) {
+      if (!tokenValueMap.containsKey(entry.getKey())) {
+        // Key is newly added to the other WorkflowToken
+        tokenValueMap.put(entry.getKey(), Lists.<NodeValueEntry>newArrayList());
+      }
+
+      // Iterate over the list of NodeValueEntry corresponding to the current key.
+      // Only add those NodeValueEntry to the merged token which already do not exist.
+      for (NodeValueEntry otherNodeValueEntry : otherTokenValueMap.get(entry.getKey())) {
+        boolean otherNodeValueEntryExist = false;
+        for (NodeValueEntry thisNodeValueEntry : tokenValueMap.get(entry.getKey())) {
+          if (thisNodeValueEntry.equals(otherNodeValueEntry)) {
+            otherNodeValueEntryExist = true;
+            break;
+          }
+        }
+        if (!otherNodeValueEntryExist) {
+          tokenValueMap.get(entry.getKey()).add(otherNodeValueEntry);
+        }
+      }
+    }
+    if (other.getMapReduceCounters() != null) {
+      setMapReduceCounters(other.getMapReduceCounters());
+    }
+  }
+
+  @Override
+  public void put(String key, String value) {
+    Preconditions.checkNotNull(nodeName, "Node name cannot be null.");
```

Comment:
Good point! I think `IllegalStateException` would be correct to throw here. Fixed.